### PR TITLE
Improve the QAM Xen testing on physical hypervisor

### PR DESCRIPTION
**Pull request by pdostal**
Hello,

I've two new testsuites - they are for Xen testing on bare metal in QAM.
The physical hypervisor is controlled by IPMI and SSH. Right now, we use standard graphical installer to install the hypervisor and then we also install guest machines using autoyast profiles.
Second testsuite will be started after the first one and this one uses SSH and the virt-manager tool and that's why it is graphical.

- Related ticket: [poo#40610](https://progress.opensuse.org/issues/40610) [poo#43739](https://progress.opensuse.org/issues/43739)
- Needles: [os-autoinst-needles-sles#1042](https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/1042)
- Verification run: [qam-regression-xen-hypervisor@12-SP3](http://pdostal-server.suse.cz/tests/827) [qam-regression-xen-hypervisor@12-SP4](http://pdostal-server.suse.cz/tests/837) [qam-regression-xen-client@12-SP3](http://pdostal-server.suse.cz/tests/834) [qam-regression-xen-client@12-SP4](http://pdostal-server.suse.cz/tests/839)

### CODE REVIEW

File: tests/virtualization/xen/dom_install.pm [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+use utils;
+
+sub run {
+    my ($self) = @_;
+    select_console 'x11';
+
+    x11_start_program('xterm');
+    send_key 'super-up';
+
+    assert_script_run "ssh root\@javor.qam.suse.cz 'zypper -n in vhostmd'";
```
Comment (ldevulder) [inline tests/virtualization/xen/dom_install.pm]: Maybe it's better to use a variable for the server name, like `QAM_XEN_SUPPORT_SERVER`, set in the test suite and do a `get_required_var('QAM_XEN_SUPPORT_SERVER');`

Also maybe `SUT_IP` can be used?

File: tests/virtualization/xen/xl_create.pm [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    select_console 'x11';
+
+    x11_start_program('xterm');
+    send_key 'super-up';
+
+    foreach my $guest (keys %xen::guests) {
+        record_info "$guest", "Starting to clone $guest to $guest-xl";
+
+        # Export the XML from virsh and convert it into Xen config file
+        assert_script_run "ssh root\@javor.qam.suse.cz 'virsh dumpxml $guest > $guest.xml'";
```
Comment (dzedro) [inline tests/virtualization/xen/xl_create.pm]: now it is javor, but later it will maybe run on another host
Variable SUT_IP could be used ?

File: lib/registration.pm [hunk 3]
```diff
@@ -680,7 +680,7 @@ sub install_docker_when_needed {
     }
     else {
         if (is_sle('<15')) {
-            assert_script_run('zypper se docker || zypper -n ar -f http://download.suse.de/ibs/SUSE:/SLE-12:/Update/standard/SUSE:SLE-12:Update.repo');
+            assert_script_run('zypper se docker || zypper -n ar -f http://download.suse.de/ibs/SUSE:/SLE-12:/Update/standard/SUSE:SLE-12:Update.repo', 120);
```
Comment (baierjan) [inline lib/registration.pm]: Is this change intentional?

File: tests/virt_autotest/login_console.pm [hunk 4]
```diff
@@ -21,13 +21,13 @@ use ipmi_backend_utils;
 
 sub login_to_console {
     my ($self, $timeout) = @_;
-    $timeout //= 120;
+    $timeout //= 600;
```
[the remark this answers is not in the document]
Comment (dzedro) [reply tests/virt_autotest/login_console.pm]: 600 was there before, it was changed few days ago to 300, in Prague we have ipmi server that have longer boot

[the remark this answers is not in the document]
Comment (dzedro) [reply tests/virt_autotest/login_console.pm]: but 600 is maybe too much

File: lib/registration.pm (hunk 3 above)
Comment (pdostal) [reply lib/registration.pm]: Ou, pardon me - it is but it should be obviously somewhere else :)

File: tests/virt_autotest/login_console.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (ldevulder) [reply tests/virt_autotest/login_console.pm]: Yes, could be good to reduce it a little. My main concern is mainly about the existing tests on IPMI backend. I don't have one to validate on my lab.

File: lib/xen.pm [hunk 5]
```diff
@@ -58,9 +63,9 @@ sub create_guest {
         # Run unattended installation for selected guest
         assert_script_run "mkdir -p /var/lib/libvirt/images/xen/";
         assert_script_run "qemu-img create -f raw /var/lib/libvirt/images/xen/$guest.raw 10G";
-        assert_script_run "virt-install --connect xen:/// --virt-type xen $extra_params --name $guest --memory 2048 --disk /var/lib/libvirt/images/xen/$guest.raw --network bridge=br0 --noautoconsole --vnc --autostart --location=$location --os-variant sles12 --wait -1 --extra-args 'autoyast=" . data_url($autoyast) . "'", timeout => 1800;
+        assert_script_run "virt-install --connect xen:/// --virt-type xen $extra_params --name $guest --memory 2048 --disk /var/lib/libvirt/images/xen/$guest.raw --network bridge=br0,mac=$macaddress --noautoconsole --vnc --autostart --location=$location --os-variant sles12 --wait -1 --extra-args 'autoyast=" . data_url($autoyast) . "'", timeout => 1800;
         # Wait for post-installation reboot as the previous command returns upon first reboot
-        script_run 'sleep 30';
+        script_run 'sleep 30', 45;
```
[the remark this answers is not in the document]
Comment (pdostal) [reply lib/xen.pm]: It is just leftover, fixed now.

File: tests/virtualization/xen/dom_install.pm (hunk 1 above)
Comment (pdostal) [reply tests/virtualization/xen/dom_install.pm]: Now I use `QAM_XEN_HYPERVISOR` variable. I still also need to rename virtuals and their DHCP / DNS records.

**Comment by pdostal** [comment]
OK, I lowered all timeouts, can you please @ldevulder check it?

The PR is now `WIP:` because of I requested to change DNS / DHCP records from `javor-$guest.qam.suse.cz` to `xen-$guest.qam.suse.cz`.

File: lib/xen.pm [hunk 6]
```diff
@@ -22,26 +22,30 @@ use utils;
...
+    #extra_params => '--paravirt',
+    #macaddress   => '52:54:00:78:73:a3',
+    #location     => 'http://mirror.suse.cz/install/SLP/SLE-15-Installer-LATEST/x86_64/DVD1/',
+    #},
+    #sles15HVM => {
+    #autoyast     => 'autoyast_xen/xen-SLES15-FV.xml',
+    #extra_params => '--hvm',
+    #macaddress   => '52:54:00:78:73:a4',
+    #location     => 'http://mirror.suse.cz/install/SLP/SLE-15-Installer-LATEST/x86_64/DVD1/',
+    #},
```
Comment (ldevulder) [inline lib/xen.pm]: Are these comments really needed? Usually we don't like commenting code in test, as git is here to track these changes. So if it doesn't hurt, please uncomment them (as I think you want to add it).

File: tests/virtualization/xen/dom_install.pm [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    x11_start_program('xterm');
+    send_key 'super-up';
+
+    assert_script_run "ssh root\@$hypervisor 'zypper -n in vhostmd'";
+
+    foreach my $guest (keys %xen::guests) {
+        record_info "$guest", "Install vm-dump-metrics on $guest-xl";
+
+        assert_script_run "ssh root\@javor-$guest.qam.suse.cz 'zypper -n in vm-dump-metrics'";
```
Comment (ldevulder) [inline tests/virtualization/xen/dom_install.pm]: This `javor-$guest.qam.suse.cz` should also be added as a variable (which should be changed for all iteration of course :)).

File: tests/virtualization/xen/install_virtmanager.pm [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+use base "x11test";
+use strict;
+use testapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+    select_console 'x11';
+    my $hypervisor = get_required_var('QAM_XEN_HYPERVISOR');
```
Comment (ldevulder) [inline tests/virtualization/xen/install_virtmanager.pm]: `$hypervisor` seems to not be used in this test, maybe better to remove it.

File: tests/virtualization/xen/patch_and_reboot.pm [hunk 9]
```diff
@@ -11,19 +11,44 @@
...
-    #    * reboot
-    #    * check that the guests are still running
+    add_test_repositories;
+    fully_patch_system;
+
+    #leave ssh console and switch to sol console
+    switch_from_ssh_to_sol_console(reset_console_flag => 'off');
+    #login
+    send_key_until_needlematch('text-login', 'ret', 360, 5);
+    #assert_screen("text-login", 1800);
```
Comment (ldevulder) [inline tests/virtualization/xen/patch_and_reboot.pm]: If it's not needed please remove this `#assert_screen [...]` line.

File: tests/virtualization/xen/virtmanager.pm [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    type_string "$hypervisor";
+    save_screenshot;    # $hypervisor written
+    send_key 'tab';
+    send_key 'spc';
+    save_screenshot;    # autoconnect ticked
+    send_key 'ret';
+    assert_screen "virt-manager_connected";
+
+
+
```
Comment (ldevulder) [inline tests/virtualization/xen/virtmanager.pm]: Why multiple blank lines?

File: lib/xen.pm (hunk 6 above)
Comment (pdostal) [reply lib/xen.pm]: I'm aware of this. Perhaps we use those domain names to communicate with VMs which are installed during the test. But those all are in our network so they should be unique as well because of at the same time the test may be running on several workers. I will discuss this next Monday in the office.

**Comment by pdostal** [comment]
@ldevulder @okurz I think I fixed everything you mentioned. Is there something else to change or can I proceed?

File: tests/virtualization/xen/xl_create.pm [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        # Change the UUID by using f00 as three first characters
+        assert_script_run "ssh root\@$hypervisor \"sed -rie 's/(uuid = \\W)(...)/\\1f00/gi' $guest.xml.cfg\"";
+        assert_script_run "ssh root\@$hypervisor 'cat $guest.xml.cfg | grep uuid'";
+
+        # Start the new VM
+        assert_script_run "ssh root\@$hypervisor xl create $guest.xml.cfg";
+        assert_script_run "ssh root\@$hypervisor xl list xl-$guest";
+
+        # Test that the new VM listens on SSH
+        assert_script_run "while true; do ssh root\@$guest.$domain hostname 2> /dev/null && break; done";
```
Comment (dzedro) [inline tests/virtualization/xen/xl_create.pm]: This can be just `ssh root\@$guest.$domain hostname` ?

Comment (ldevulder) [reply tests/virtualization/xen/xl_create.pm]: Oh or maybe it's a way to verify that's the VM is started/running, right?

Comment (dzedro) [reply tests/virtualization/xen/xl_create.pm]: It's kind of retry, the VM can be in that time unreachable so ok, can be.

**Comment by dzedro** [comment]
roll the dice
